driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java

```diff
@@ -89,40 +89,48 @@ static MongoException createSpecialWriteConcernException(final ResponseBuffers r
     }
 
     static BsonTimestamp getOperationTime(final ResponseBuffers responseBuffers) {
-        try {
-            BsonValue operationTime = getField(createBsonReader(responseBuffers), "operationTime");
-            if (operationTime == null) {
-                return null;
-            }
-            return operationTime.asTimestamp();
-        } finally {
-            responseBuffers.reset();
-        }
+        return getFieldValueAsTimestamp(responseBuffers, "operationTime");
     }
 
     static BsonDocument getClusterTime(final ResponseBuffers responseBuffers) {
-        return getFieldDocument(responseBuffers, "$clusterTime");
+        return getFieldValueAsDocument(responseBuffers, "$clusterTime");
     }
 
-    static BsonDocument getClusterTime(final BsonDocument response) {
-        BsonValue clusterTime = response.get("$clusterTime");
-        if (clusterTime == null) {
-            return null;
+    static BsonTimestamp getSnapshotTimestamp(final ResponseBuffers responseBuffers) {
+        BsonValue atClusterTimeValue = getNestedFieldValue(responseBuffers, "cursor", "atClusterTime");
+        if (atClusterTimeValue == null) {
+            atClusterTimeValue = getFieldValue(responseBuffers, "atClusterTime");
+        }
+        if (atClusterTimeValue != null && atClusterTimeValue.getBsonType() == BsonType.TIMESTAMP) {
+            return atClusterTimeValue.asTimestamp();
```

Comment:
I would expect that if `atClusterTimeValue.isTimestamp()` is true, then `atClusterTimeValue.asTimestamp()` can be called safely. If this is the case, maybe we should replace `atClusterTimeValue.getBsonType() == BsonType.TIMESTAMP` with ``atClusterTimeValue.isTimestamp()`?

Reply:
Done